cmd/clusterctl/client/cluster/topology_test.go

```diff
@@ -0,0 +1,692 @@
+    joinConfiguration:
+      discovery: {}
+      nodeRegistration:
+        criSocket: /var/run/containerd/containerd.sock
+        kubeletExtraArgs:
+          cgroup-driver: cgroupfs
+          eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
+  machineTemplate:
+    infrastructureRef:
+      apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+      kind: DockerMachineTemplate
+      name: my-cluster-control-plane-44cd4
+      namespace: default
+    metadata:
+      labels:
+        cluster.x-k8s.io/cluster-name: my-cluster
+        topology.cluster.x-k8s.io/owned: ""
+    nodeDrainTimeout: 1s
+  replicas: 1
+  rolloutStrategy:
+    rollingUpdate:
+      maxSurge: 1
+    type: RollingUpdate
+  version: v1.21.2
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: DockerMachineTemplate
+metadata:
+  annotations:
+    cluster.x-k8s.io/cloned-from-groupkind: DockerMachineTemplate.infrastructure.cluster.x-k8s.io
+    cluster.x-k8s.io/cloned-from-name: control-plane
+  creationTimestamp: "2021-12-29T09:45:03Z"
+  generation: 1
+  labels:
+    cluster.x-k8s.io/cluster-name: my-cluster
+    topology.cluster.x-k8s.io/owned: ""
+  name: my-cluster-control-plane-44cd4
+  namespace: default
+  ownerReferences:
+    - apiVersion: cluster.x-k8s.io/v1beta1
+      kind: Cluster
+      name: my-cluster
+      uid: 3ba5ce4f-d279-4edb-8ade-62a2381d11a8
+  resourceVersion: "21033"
+  uid: e70775ca-f704-4ae0-94a5-d84f7b4b94f5
+spec:
+  template:
+    spec:
+      extraMounts:
+        - containerPath: /var/run/docker.sock
+          hostPath: /var/run/docker.sock
+      preLoadImages:
+        - gcr.io/kakaraparthy-devel/kindest/kindnetd:0.5.4
+`)
+
+// modifiedClusterYAML changes the control plane replicas from 1 to 3.
+var modifiedClusterYAML = []byte(`
+---
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: Cluster
+metadata:
+  name: "my-cluster"
+  namespace: default
+  labels:
+    cni: kindnet
+spec:
+  clusterNetwork:
+    services:
+      cidrBlocks: ["10.128.0.0/12"]
+    pods:
+      cidrBlocks: ["192.168.0.0/16"]
+    serviceDomain: "cluster.local"
+  topology:
+    class: my-cluster-class
+    version: v1.21.2
+    controlPlane:
+      metadata: {}
+      replicas: 3
+`)
+
+func Test_topologyClient_DryRun(t *testing.T) {
+	type args struct {
+		in *DryRunInput
+	}
+	type item struct {
+		kind       string
+		namespace  string
+		namePrefix string
+	}
+	type out struct {
+		affectedClusters       []client.ObjectKey
+		affectedClusterClasses []client.ObjectKey
+		created                []item
+		modified               []item
+		deleted                []item
+	}
+	tests := []struct {
+		name            string
+		existingObjects []*unstructured.Unstructured
+		args            args
+		want            out
+		wantErr         bool
+	}{
+		{
+			name: "Input with new ClusterClass and new Cluster",
+			args: args{
+				in: func() *DryRunInput {
+					objs, err := utilyaml.ToUnstructured(newClusterClassAndClusterYAML)
```

Comment:
This is the first implementation of the tests. May be there is scope here to use the builder library to generate the test data.